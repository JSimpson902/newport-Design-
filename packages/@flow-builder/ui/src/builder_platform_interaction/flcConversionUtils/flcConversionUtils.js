import { ElementType } from 'builder_platform_interaction/flowUtils';
import { ELEMENT_TYPE, CONNECTOR_TYPE } from 'builder_platform_interaction/flowMetadata';
import { addElement } from 'builder_platform_interaction/actions';
import { createEndElement, createConnector } from 'builder_platform_interaction/elementFactory';
import { getConfigForElementType } from 'builder_platform_interaction/elementConfig';
import {
    supportsChildren,
    linkBranch,
    addElementToState,
    isRootOrEndElement,
    linkElement
} from 'builder_platform_interaction/flcBuilderUtils';

// TODO: FLC Hack: Temp magic number to identify an flc flow
const LOCATION_Y_FLC_FLOW = 500;

// TODO: FLC Hack: Magic number to support identifying merge elements
const LOCATION_X_MERGE_MARKER = 666;

const flcExtraProps = ['next', 'prev', 'children', 'parent', 'childIndex', 'isTerminal'];

export function isFixedLayoutCanvas(startElement) {
    return startElement.locationY === LOCATION_Y_FLC_FLOW;
}

function getChildReferencesKey(parentElement) {
    return getConfigForElementType(parentElement.elementType).childReferenceKey;
}

function createConnectorHelper({ source, childSource, target, label = '', type = CONNECTOR_TYPE.REGULAR }) {
    return createConnector(source, childSource, target, label, type);
}

/**
 * Find the childSource at a given index for a parentElement
 * @param {Object} parentElement
 * @param {number} index
 */
function findChildSource(parentElement, index) {
    const { singular, plural } = getChildReferencesKey(parentElement);
    return parentElement[plural][index - 1][singular];
}

/**
 * Find the index for a childSource guid
 * @param {Object} element
 * @param {*} guid
 */
function findConnectionIndex(parentElement, guid) {
    const { singular, plural } = getChildReferencesKey(parentElement);
    const index = parentElement[plural].findIndex(entry => entry[singular] === guid);

    return index === -1 ? 0 : index + 1;
}

function createElementHelper(elementType, guid) {
    return {
        elementType,
        guid,
        label: elementType,
        value: elementType,
        text: elementType,
        name: elementType,
        prev: null,
        next: null
    };
}

/**
 * Creates a root element and links it with the start element
 * @param {string} startElementGuid
 */
const createRootElement = (startElementGuid = null) => {
    return {
        ...createElementHelper(ELEMENT_TYPE.ROOT_ELEMENT, ElementType.ROOT),
        children: [startElementGuid]
    };
};

/**
 * Adds a null'ed children array to a parentElement
 * @param {Object} element
 */
export function initializeChildren(element) {
    const { childReferenceKey } = getConfigForElementType(element.elementType);
    // TODO: FLC find better way to determine child count (there could be more to account for faults)
    element.children = element.children || new Array(element[childReferenceKey.plural].length + 1).fill(null);
}

/**
 * Find the start element
 * @param {Object} elements - map of guid -> element
 * @return the start element
 */
function findStartElement(elements) {
    return Object.values(elements).find(ele => ele.elementType === ELEMENT_TYPE.START_ELEMENT);
}

/**
 * Converts an FFC ui model to a FLC ui model by augmenting the FFC ui model
 * with the extra properties needed by the FLC ui model
 * @param {Object} ffcUiModel - an ffc ui model
 * @return a flc ui model
 */
export function convertToFlc(ffcUiModel) {
    const { connectors, elements } = ffcUiModel;

    connectors.forEach(connector => {
        const { source, target, childSource } = connector;
        const sourceElement = elements[source];
        const targetElement = elements[target];

        if (supportsChildren(sourceElement) && !isMergeElement(targetElement)) {
            initializeChildren(sourceElement);
            const childIndex = findConnectionIndex(sourceElement, childSource);
            linkBranch(elements, sourceElement, childIndex, targetElement);
        } else {
            targetElement.prev = source;
            sourceElement.next = target;
        }
    });

    const startElement = findStartElement(elements);
    const rootElement = createRootElement();
    addElementToState(rootElement, elements);
    linkBranch(elements, rootElement, 0, startElement);
    fixFlcProperties(ffcUiModel, startElement, rootElement);

    return { elements, canvasElements: [], connectors: [] };
}

/**
 * Adds a root and end element for a new flow
 * @param {Object} storeInstance - the storeInstance
 * @param {string} startElementGuid - the start element guid
 */
export function addRootAndEndElements(storeInstance, startElementGuid) {
    const endElement = createEndElement({ prev: startElementGuid });
    const rootElement = createRootElement(startElementGuid);

    storeInstance.dispatch(addElement(endElement));
    storeInstance.dispatch(addElement(rootElement));
}

/**
 * Creates the connectors for the a parent's children elements until they merge back
 * @return whether any connector was created
 */
function createConnectorsForChildren(newElements, newConnectors, newCanvasElements, parentElement) {
    let createdConnector = false;

    parentElement.children.forEach((child, i) => {
        if (child) {
            const nextElement = newElements[child];

            // create the connector to the head of the branch
            if (nextElement && nextElement.elementType !== ELEMENT_TYPE.END_ELEMENT) {
                let type, childSource;

                if (i === 0) {
                    type = CONNECTOR_TYPE.DEFAULT;
                    childSource = null;
                } else {
                    type = CONNECTOR_TYPE.REGULAR;
                    childSource = findChildSource(parentElement, i);
                }
                const source = parentElement.guid;
                const target = nextElement.guid;
                const label = '';

                newConnectors.push(createConnectorHelper({ source, childSource, target, label, type }));
                createdConnector = true;
            }
        }

        // branch connectors for the child's successors until they merge back
        if (child != null) {
            convertFromFlcHelper(newElements, newConnectors, newCanvasElements, newElements[child], parentElement);
        }
    });

    return createdConnector;
}

// creates the connector for an element and its successors
function convertFromFlcHelper(newElements, newConnectors, newCanvasElements, element, parentElement) {
    let prevElement = null;
    let isChildless;

    while (element) {
        if (element.elementType !== ELEMENT_TYPE.END_ELEMENT) {
            const newElement = { ...element };

            const elementSupportsChildren = supportsChildren(element);
            if (elementSupportsChildren) {
                // Creates the connectors for the children
                // If not connector was created (because no children), then we need to
                // create a connector for the parent to its next element
                isChildless = !createConnectorsForChildren(newElements, newConnectors, newCanvasElements, element);
            } else {
                isChildless = false;
            }

            const nextElement = newElements[element.next];
            if (nextElement) {
                if (nextElement.elementType !== ELEMENT_TYPE.END_ELEMENT) {
                    if (elementSupportsChildren) {
                        nextElement.locationX = LOCATION_X_MERGE_MARKER;
                    }

                    // create a connector to the next element for non-branching elements,
                    // or childless branching elements
                    if (!elementSupportsChildren || isChildless) {
                        newConnectors.push(
                            createConnectorHelper({
                                source: element.guid,
                                target: element.next,
                                type: isChildless ? CONNECTOR_TYPE.DEFAULT : CONNECTOR_TYPE.REGULAR
                            })
                        );
                    }
                }
            } else if (parentElement) {
                // we are on a branch and don't have a next element, this means the parentElement's next
                // is where the flow execution continues to, so create a connector
                if (parentElement.next && newElements[parentElement.next].elementType !== ELEMENT_TYPE.END_ELEMENT) {
                    newConnectors.push(
                        createConnectorHelper({
                            source: element.guid,
                            target: parentElement.next
                        })
                    );
                }
            }

            // delete extra flc properties
            flcExtraProps.forEach(prop => delete newElement[prop]);

            addElementToState(newElement, newElements, newCanvasElements);
        }

        // the prevElement supports children, this means element is a merge element, so mark it so
        if (prevElement && supportsChildren(prevElement)) {
            element.locationX = LOCATION_X_MERGE_MARKER;
        }

        prevElement = element;
        element = newElements[element.next];
    }
}

/**
 * Converts an FLC ui model to a FFC ui model by striping the extra FLC ui model properties
 * @param {Object} uiModel - an FLC ui model
 * @return {Object} an FFC ui model
 */
export function convertFromFlc(uiModel) {
    const { elements } = uiModel;

    const newConnectors = [];
    const newCanvasElements = [];

    // clone all elements
    let newElements = Object.values(elements).reduce((acc, element) => {
        acc[element.guid] = { ...element };
        return acc;
    }, {});

    const startElement = findStartElement(newElements);
    if (startElement) {
        startElement.locationY = LOCATION_Y_FLC_FLOW;

        convertFromFlcHelper(newElements, newConnectors, newCanvasElements, startElement);

        // filter out root and end elements
        newElements = Object.values(newElements)
            .filter(element => !isRootOrEndElement(element))
            .reduce((acc, element) => {
                acc[element.guid] = element;
                return acc;
            }, {});
    }

    return { ...uiModel, elements: newElements, connectors: newConnectors, canvasElements: newCanvasElements };
}

function isMergeElement(element) {
    return element && element.locationX === LOCATION_X_MERGE_MARKER;
}

function fixFlcPropertiesHelper(storeState, element, visited) {
    const { elements } = storeState;
    initializeChildren(element);
    element.children.forEach(child => fixFlcProperties(storeState, elements[child], element, visited));
    const nextElement = elements[element.next];
    if (isMergeElement(nextElement)) {
        // clear merge marker now that the branch element has its next pointer computed
        nextElement.locationX = 0;
    }
}

// creates END elements and fixes merging branches pointers
function fixFlcProperties(storeState, element, parentElement, visited = {}) {
    const { elements } = storeState;
    let prevElement;
    const firstElement = element;

    // check that we have an element and it hasn't been visited
    while (element && !visited[element.guid]) {
        visited[element.guid] = true;

        // clear any magic numbers
        Object.assign(element, { locationX: 0, locationY: 0 });

        if (supportsChildren(element)) {
            fixFlcPropertiesHelper(storeState, element, visited);
        }

        prevElement = element;
        element = elements[element.next];
        if (isMergeElement(element)) {
            // if we hit a merge marker, and have a parentElement, that means that prevElement is the last element of a branch
            if (parentElement) {
                prevElement.next = null;
                parentElement.next = element.guid;
                linkElement(elements, parentElement);

                // terminate the loop since we just hit the last branch element
                element = null;
            }
        } else if (!element) {
            const endElement = createEndElement({ prev: prevElement.guid });
            linkElement(elements, endElement);

            // mark the branch as having an end element
            firstElement.isTerminal = true;
        }
    }
}

/**
 * Stringify Store State for debugging
 * @param {Object} storeState
 */
export function prettyPrintStoreState(storeState) {
    if (storeState) {
        const { elements, canvasElements, connectors } = storeState;

        // eslint-disable-next-line
        console.log(JSON.parse(JSON.stringify({ elements, connectors, canvasElements })));

        let res = (canvasElements || []).map(guid => {
            const { next, prev, children, parent, childIndex, elementType } = elements[guid];
            const label = elements[guid].label || guid;
            let nextLabel = next ? elements[next] : null;
            let prevLabel = prev ? elements[prev] : null;
            let parentLabel = parent ? elements[parent] : null;

            nextLabel = nextLabel ? nextLabel.label || nextLabel.guid : next;
            prevLabel = prevLabel ? prevLabel.label || prevLabel.guid : prev;
            parentLabel = parentLabel ? parentLabel.label || parentLabel.guid : parent;

            return `${label}: guid: ${guid} elementType: ${elementType} next: ${nextLabel}, prev: ${prevLabel}, children: ${children}, parent: ${parentLabel}, childIndex: ${childIndex}`;
        });

        res = res.concat(
            (connectors || []).map(
                ({ source, target }) => `source: ${elements[source].label}, target: ${elements[target].label}`
            )
        );

        // eslint-disable-next-line
        console.log(res.join('\n'));
    }
}
