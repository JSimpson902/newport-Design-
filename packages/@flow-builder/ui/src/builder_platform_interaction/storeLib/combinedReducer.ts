// @ts-nocheck
/**
 * Combine multiple reducer to a one reducer function.
 * @param {Object} reducers object containing key value pair of all the reducers. Key is name of the property of the state of a store on which reduction needs to happen.
 * @returns {Function} one reducer function.
 */
export function combinedReducer(reducers) {
    return function combined(state = {}, action) {
        let updated = false;
        const newState = {};
        const reducerKeys = Object.keys(reducers);
        reducerKeys.forEach(key => {
            newState[key] = reducers[key](state[key], action);
            updated = updated || newState[key] !== state[key];
        });
        return updated ? newState : state;
    };
}
