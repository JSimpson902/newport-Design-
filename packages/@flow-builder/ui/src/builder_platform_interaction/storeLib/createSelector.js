import { memoize } from 'builder_platform_interaction/commonUtils';

/**
 * This is used to derive data from a state in the store.
 * @param {Function[]} selectors An array of selectors that are needed for transformation.
 * @param {Function} transformation A function, which derives data based on selectors.
 * @param {boolean} [false] shouldMemoize Set to true to memoize the selectors and the transformation.
 * @returns {Function} A function in which a user passes the state and it transforms an appropriate piece of data.
 */
export function createSelector(selectors, transformation, shouldMemoize) {
    if (!selectors || !selectors.length || !Array.isArray(selectors)) {
        throw new Error(`could not transform the ${selectors}.`);
    }

    let transformationFunc = transformation;
    // Memoize the transformation, if requested
    if (shouldMemoize) {
        transformationFunc = memoize(function() {
            return transformation.apply(null, arguments);
        });
    }

    let selector = function() {
        // Invoke each selector with supplied arguments, collecting results into an array.
        const dataFromSelectors = selectors.map(arg => arg.apply(null, arguments));
        // Invoke the transformation, passing it the array of selectors' results.
        return transformationFunc(...dataFromSelectors);
    };

    // Memoize the selector, if requested
    if (shouldMemoize) {
        selector = memoize(selector);
    }

    return selector;
}
