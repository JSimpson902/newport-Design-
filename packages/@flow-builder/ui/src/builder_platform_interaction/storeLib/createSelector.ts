// @ts-nocheck
import { commonUtils } from 'builder_platform_interaction/sharedUtils';
const { memoize } = commonUtils;

/**
 * This is used to derive data from a state in the store.
 *
 * @param {Function[]} selectors An array of selectors that are needed for transformation.
 * @param {Function} transformation A function, which derives data based on selectors.
 * @param {boolean} shouldMemoize Set to true to memoize the selectors and the transformation.
 * @returns {Function} A function in which a user passes the state and it transforms an appropriate piece of data.
 */
export function createSelector(selectors, transformation, shouldMemoize = false) {
    if (!selectors || !selectors.length || !Array.isArray(selectors)) {
        throw new Error(`could not transform the ${selectors}.`);
    }

    let transformationFunc = transformation;
    // Memoize the transformation, if requested
    if (shouldMemoize) {
        transformationFunc = memoize(function () {
            return transformation.apply(null, arguments);
        });
    }

    let selector = function (...args: any[]) {
        // Invoke each selector with supplied arguments, collecting results into an array.
        const dataFromSelectors = selectors.map((arg) => arg.apply(null, args));
        // Invoke the transformation, passing it the array of selectors' results.
        return transformationFunc(...dataFromSelectors);
    };

    // Memoize the selector, if requested
    if (shouldMemoize) {
        selector = memoize(selector);
    }

    return selector;
}
