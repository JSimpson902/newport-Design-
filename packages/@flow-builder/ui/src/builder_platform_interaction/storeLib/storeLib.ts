// @ts-nocheck
export { combinedReducer } from './combinedReducer';
export { createSelector } from './createSelector';
export { Store } from './store';
export { generateGuid } from './guidGenerator';
export { deepCopy } from './deepCopy';
export { isPlainObject } from './isPlainObject';
