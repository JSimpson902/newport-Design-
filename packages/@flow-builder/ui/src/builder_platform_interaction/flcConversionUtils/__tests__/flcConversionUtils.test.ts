// @ts-nocheck
import { deepCopy } from 'builder_platform_interaction/storeLib';

import sanity from './flcUiModels/sanity';
import oneScreen from './flcUiModels/one-screen';
import oneElementWithFault from './flcUiModels/one-element-with-fault';
import decisionOneChildOnEachBranchNextIsNotEnd from './flcUiModels/decision-one-child-on-each-branch-next-is-not-end';

import { convertFromFlc, convertToFlc, canConvertToFlc } from '../flcConversionUtils';

jest.mock('builder_platform_interaction/elementFactory', () => {
    const flowMetadata = require('builder_platform_interaction/flowMetadata');
    return {
        createConnector: (source, childSource, target, label, type, isSelected = false) => {
            return {
                guid: `${source} -> ${target}`,
                source,
                childSource,
                target,
                label,
                type,
                config: { isSelected }
            };
        },
        createEndElement: ({ prev }) => {
            return {
                guid: `end-element-guid (${prev})`,
                elementType: flowMetadata.ELEMENT_TYPE.END_ELEMENT,
                prev,
                next: null
            };
        }
    };
});

jest.mock('builder_platform_interaction/contextLib', () => {
    return require('builder_platform_interaction_mocks/contextLib');
});

jest.mock('builder_platform_interaction/storeLib', () => {
    const actual = require('builder_platform_interaction_mocks/storeLib');

    return {
        generateGuid: jest.fn().mockImplementation(() => {
            return 'end-element-guid';
        }),
        deepCopy: actual.deepCopy
    };
});

function sortCanvasElementsAndConnectors({ elements, canvasElements, connectors }) {
    canvasElements.sort();
    connectors = connectors
        .sort((a, b) => b.guid.localeCompare(a.guid))
        .map(connector => {
            if (connector.childSource == null) {
                delete connector.childSource;
            }
            return connector;
        });
    return { elements, canvasElements, connectors };
}

function testConversions(flcUiModel) {
    let ffcUiModel;
    it('from flc', () => {
        ffcUiModel = sortCanvasElementsAndConnectors(convertFromFlc(deepCopy(flcUiModel)));
        expect(ffcUiModel).toMatchSnapshot();
    });

    it('to flc', () => {
        expect(convertToFlc(deepCopy(ffcUiModel))).toEqual(flcUiModel);
    });
}

describe('converts', () => {
    describe('sanity', () => {
        testConversions(sanity);
    });

    describe('one-screen', () => {
        testConversions(oneScreen);
    });

    describe('one decision with one child on each branch, followed by a screen element', () => {
        testConversions(decisionOneChildOnEachBranchNextIsNotEnd);
    });

    describe('one element with fault', () => {
        testConversions(oneElementWithFault);
    });

    describe('can convert to flc', () => {
        describe('empty flow is valid', () => {
            const startGuid = 'start';
            const connectors = [{ source: 'start', target: null }];
            expect(canConvertToFlc(startGuid, connectors)).toBeTruthy();
        });

        describe('if flow is valid', () => {
            const startGuid = 'start';
            const connectors = [
                { source: 'start', target: 'if' },
                { source: 'if', target: 'n1' },
                { source: 'if', target: 'n2' },
                { source: 'n1', target: 'merge' },
                { source: 'n2', target: 'merge' }
            ];
            expect(canConvertToFlc(startGuid, connectors)).toBeTruthy();
        });

        // describe('loop flow is valid', () => {
        //     const startGuid = 'start';
        //     const connectors = [
        //         { source: 'start', target: 'loop' },
        //         { source: 'loop', target: 'n1' },
        //         { source: 'loop', target: 'n2' },
        //         { source: 'n1', target: 'loop' }
        //     ];
        //     expect(canConvertToFlc(startGuid, connectors)).toBeTruthy();
        // });

        describe('cross if flow invalid', () => {
            const startGuid = 'start';
            const connectors = [
                { source: 'start', target: 'if' },
                { source: 'if', target: 'n1' },
                { source: 'if', target: 'n2' },
                { source: 'n1', target: 'n2' },
                { source: 'n1', target: 'merge' },
                { source: 'n2', target: 'merge' }
            ];
            expect(canConvertToFlc(startGuid, connectors)).toBeFalsy();
        });

        describe('loop back before if flow is invalid', () => {
            const startGuid = 'start';
            const connectors = [
                { source: 'start', target: 'if' },
                { source: 'if', target: 'n1' },
                { source: 'if', target: 'n2' },
                { source: 'n1', target: 'start' }
            ];
            expect(canConvertToFlc(startGuid, connectors)).toBeFalsy();
        });

        describe('loop back to if flow is invalid', () => {
            const startGuid = 'start';
            const connectors = [
                { source: 'start', target: 'if' },
                { source: 'if', target: 'n1' },
                { source: 'if', target: 'n2' },
                { source: 'n1', target: 'if' }
            ];
            expect(canConvertToFlc(startGuid, connectors)).toBeFalsy();
        });

        describe('jump out of if flow is invalid', () => {
            const startGuid = 'start';
            const connectors = [
                { source: 'start', target: 'if' },
                { source: 'if', target: 'n1' },
                { source: 'if', target: 'n2' },
                { source: 'n1', target: 'merge' },
                { source: 'n2', target: 'merge' },
                { source: 'merge', target: 'n3' },
                { source: 'n1', target: 'n3' }
            ];
            expect(canConvertToFlc(startGuid, connectors)).toBeFalsy();
        });
    });
});
