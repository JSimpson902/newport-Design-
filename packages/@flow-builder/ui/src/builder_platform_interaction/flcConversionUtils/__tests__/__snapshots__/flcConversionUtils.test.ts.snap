// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts one decision with one child on each branch, followed by a screen element from flc 1`] = `
Object {
  "canvasElements": Array [
    "decision-element-guid",
    "screen-after-decision-element-guid",
    "screen-left-element-guid",
    "screen-right-element-guid",
    "start-element-guid",
  ],
  "connectors": Array [
    Object {
      "config": Object {
        "isSelected": false,
      },
      "guid": "start-element-guid -> decision-element-guid",
      "label": "",
      "source": "start-element-guid",
      "target": "decision-element-guid",
      "type": "REGULAR",
    },
    Object {
      "config": Object {
        "isSelected": false,
      },
      "guid": "screen-right-element-guid -> screen-after-decision-element-guid",
      "label": "",
      "source": "screen-right-element-guid",
      "target": "screen-after-decision-element-guid",
      "type": "REGULAR",
    },
    Object {
      "config": Object {
        "isSelected": false,
      },
      "guid": "screen-left-element-guid -> screen-after-decision-element-guid",
      "label": "",
      "source": "screen-left-element-guid",
      "target": "screen-after-decision-element-guid",
      "type": "REGULAR",
    },
    Object {
      "childSource": "outcome-element-guid",
      "config": Object {
        "isSelected": false,
      },
      "guid": "decision-element-guid -> screen-right-element-guid",
      "label": "",
      "source": "decision-element-guid",
      "target": "screen-right-element-guid",
      "type": "REGULAR",
    },
    Object {
      "config": Object {
        "isSelected": false,
      },
      "guid": "decision-element-guid -> screen-left-element-guid",
      "label": "",
      "source": "decision-element-guid",
      "target": "screen-left-element-guid",
      "type": "DEFAULT",
    },
  ],
  "elements": Object {
    "decision-element-guid": Object {
      "availableConnections": Array [
        Object {
          "childReference": "outcome-element-guid",
          "type": "REGULAR",
        },
        Object {
          "type": "DEFAULT",
        },
      ],
      "childReferences": Array [
        Object {
          "childReference": "outcome-element-guid",
        },
      ],
      "config": Object {
        "isHighlighted": false,
        "isSelectable": true,
        "isSelected": false,
      },
      "connectorCount": 0,
      "defaultConnectorLabel": "Default Outcome",
      "description": "",
      "elementType": "Decision",
      "guid": "decision-element-guid",
      "isCanvasElement": true,
      "label": "d1",
      "locationX": 625,
      "locationY": 240,
      "maxConnections": 2,
      "name": "d1",
    },
    "outcome-element-guid": Object {
      "conditionLogic": "and",
      "conditions": Array [
        Object {
          "leftHandSide": "$Flow.CurrentRecord",
          "operator": "EqualTo",
          "rightHandSide": "a",
          "rightHandSideDataType": "String",
          "rowIndex": "cbf949b8-cb9c-4f36-931f-1b128912360d",
        },
      ],
      "dataType": "Boolean",
      "elementType": "OUTCOME",
      "guid": "outcome-element-guid",
      "label": "d1out",
      "name": "d1out",
    },
    "screen-after-decision-element-guid": Object {
      "config": Object {
        "isHighlighted": false,
        "isSelectable": true,
        "isSelected": false,
      },
      "connectorCount": 0,
      "description": "",
      "elementType": "Screen",
      "filterType": "all",
      "filters": Array [
        Object {
          "leftHandSide": "",
          "operator": "",
          "rightHandSide": "",
          "rightHandSideDataType": "",
          "rowIndex": "cda83bda-2840-470c-aa57-1a8cfedee523",
        },
      ],
      "guid": "screen-after-decision-element-guid",
      "isCanvasElement": true,
      "locationX": 666,
      "locationY": 576,
      "maxConnections": 1,
      "object": "",
      "objectIndex": "12c23e8e-4104-40ba-86e0-feb73e787792",
      "triggerType": "None",
    },
    "screen-left-element-guid": Object {
      "config": Object {
        "isHighlighted": false,
        "isSelectable": true,
        "isSelected": false,
      },
      "connectorCount": 0,
      "description": "",
      "elementType": "Screen",
      "filterType": "all",
      "filters": Array [
        Object {
          "leftHandSide": "",
          "operator": "",
          "rightHandSide": "",
          "rightHandSideDataType": "",
          "rowIndex": "cda83bda-2840-470c-aa57-1a8cfedee523",
        },
      ],
      "guid": "screen-left-element-guid",
      "isCanvasElement": true,
      "locationX": 493,
      "locationY": 360,
      "maxConnections": 1,
      "object": "",
      "objectIndex": "12c23e8e-4104-40ba-86e0-feb73e787792",
      "triggerType": "None",
    },
    "screen-right-element-guid": Object {
      "config": Object {
        "isHighlighted": false,
        "isSelectable": true,
        "isSelected": false,
      },
      "connectorCount": 0,
      "description": "",
      "elementType": "Screen",
      "filterType": "all",
      "filters": Array [
        Object {
          "leftHandSide": "",
          "operator": "",
          "rightHandSide": "",
          "rightHandSideDataType": "",
          "rowIndex": "cda83bda-2840-470c-aa57-1a8cfedee523",
        },
      ],
      "guid": "screen-right-element-guid",
      "isCanvasElement": true,
      "locationX": 757,
      "locationY": 360,
      "maxConnections": 1,
      "object": "",
      "objectIndex": "12c23e8e-4104-40ba-86e0-feb73e787792",
      "triggerType": "None",
    },
    "start-element-guid": Object {
      "config": Object {
        "isHighlighted": false,
        "isSelectable": true,
        "isSelected": false,
      },
      "connectorCount": 0,
      "description": "",
      "elementType": "START_ELEMENT",
      "filterType": "all",
      "filters": Array [
        Object {
          "leftHandSide": "",
          "operator": "",
          "rightHandSide": "",
          "rightHandSideDataType": "",
          "rowIndex": "e7f9f0cf-3953-43b3-a578-3ddc8a10e4df",
        },
      ],
      "guid": "start-element-guid",
      "isCanvasElement": true,
      "locationX": 500,
      "locationY": 120,
      "maxConnections": 1,
      "object": "",
      "objectIndex": "2832cafc-39fc-41bf-95b1-c6d3881f6705",
      "triggerType": "None",
    },
  },
}
`;

exports[`converts one element with fault from flc 1`] = `
Object {
  "canvasElements": Array [
    "fault-element-guid",
    "record-create-element-guid",
    "start-element-guid",
  ],
  "connectors": Array [
    Object {
      "config": Object {
        "isSelected": false,
      },
      "guid": "start-element-guid -> record-create-element-guid",
      "label": "",
      "source": "start-element-guid",
      "target": "record-create-element-guid",
      "type": "REGULAR",
    },
    Object {
      "config": Object {
        "isSelected": false,
      },
      "guid": "record-create-element-guid -> fault-element-guid",
      "label": "",
      "source": "record-create-element-guid",
      "target": "fault-element-guid",
      "type": "FAULT",
    },
  ],
  "elements": Object {
    "fault-element-guid": Object {
      "config": Object {
        "isHighlighted": false,
        "isSelectable": true,
        "isSelected": false,
      },
      "connectorCount": 0,
      "description": "",
      "elementType": "RecordCreate",
      "filterType": "all",
      "filters": Array [
        Object {
          "leftHandSide": "",
          "operator": "",
          "rightHandSide": "",
          "rightHandSideDataType": "",
          "rowIndex": "dda83bda-2840-470c-aa57-1a8cfedee523",
        },
      ],
      "guid": "fault-element-guid",
      "isCanvasElement": true,
      "locationX": 889,
      "locationY": 360,
      "maxConnections": 1,
      "object": "",
      "objectIndex": "92c23e8e-4104-40ba-86e0-feb73e787792",
      "triggerType": "None",
    },
    "record-create-element-guid": Object {
      "config": Object {
        "isHighlighted": false,
        "isSelectable": true,
        "isSelected": false,
      },
      "connectorCount": 0,
      "description": "",
      "elementType": "RecordCreate",
      "filterType": "all",
      "filters": Array [
        Object {
          "leftHandSide": "",
          "operator": "",
          "rightHandSide": "",
          "rightHandSideDataType": "",
          "rowIndex": "cda83bda-2840-470c-aa57-1a8cfedee523",
        },
      ],
      "guid": "record-create-element-guid",
      "isCanvasElement": true,
      "locationX": 625,
      "locationY": 240,
      "maxConnections": 1,
      "object": "",
      "objectIndex": "12c23e8e-4104-40ba-86e0-feb73e787792",
      "triggerType": "None",
    },
    "start-element-guid": Object {
      "config": Object {
        "isHighlighted": false,
        "isSelectable": true,
        "isSelected": false,
      },
      "connectorCount": 0,
      "description": "",
      "elementType": "START_ELEMENT",
      "filterType": "all",
      "filters": Array [
        Object {
          "leftHandSide": "",
          "operator": "",
          "rightHandSide": "",
          "rightHandSideDataType": "",
          "rowIndex": "cda83bda-2840-470c-aa57-1a8cfedee523",
        },
      ],
      "guid": "start-element-guid",
      "isCanvasElement": true,
      "locationX": 500,
      "locationY": 120,
      "maxConnections": 1,
      "object": "",
      "objectIndex": "12c23e8e-4104-40ba-86e0-feb73e787792",
      "triggerType": "None",
    },
  },
}
`;

exports[`converts one-screen from flc 1`] = `
Object {
  "canvasElements": Array [
    "screen-element-guid",
    "start-element-guid",
  ],
  "connectors": Array [
    Object {
      "config": Object {
        "isSelected": false,
      },
      "guid": "start-element-guid -> screen-element-guid",
      "label": "",
      "source": "start-element-guid",
      "target": "screen-element-guid",
      "type": "REGULAR",
    },
  ],
  "elements": Object {
    "screen-element-guid": Object {
      "config": Object {
        "isHighlighted": false,
        "isSelectable": true,
        "isSelected": false,
      },
      "connectorCount": 0,
      "description": "",
      "elementType": "Screen",
      "filterType": "all",
      "filters": Array [
        Object {
          "leftHandSide": "",
          "operator": "",
          "rightHandSide": "",
          "rightHandSideDataType": "",
          "rowIndex": "cda83bda-2840-470c-aa57-1a8cfedee523",
        },
      ],
      "guid": "screen-element-guid",
      "isCanvasElement": true,
      "locationX": 625,
      "locationY": 240,
      "maxConnections": 1,
      "object": "",
      "objectIndex": "12c23e8e-4104-40ba-86e0-feb73e787792",
      "triggerType": "None",
    },
    "start-element-guid": Object {
      "config": Object {
        "isHighlighted": false,
        "isSelectable": true,
        "isSelected": false,
      },
      "connectorCount": 0,
      "description": "",
      "elementType": "START_ELEMENT",
      "filterType": "all",
      "filters": Array [
        Object {
          "leftHandSide": "",
          "operator": "",
          "rightHandSide": "",
          "rightHandSideDataType": "",
          "rowIndex": "cda83bda-2840-470c-aa57-1a8cfedee523",
        },
      ],
      "guid": "start-element-guid",
      "isCanvasElement": true,
      "locationX": 500,
      "locationY": 120,
      "maxConnections": 1,
      "object": "",
      "objectIndex": "12c23e8e-4104-40ba-86e0-feb73e787792",
      "triggerType": "None",
    },
  },
}
`;

exports[`converts sanity from flc 1`] = `
Object {
  "canvasElements": Array [
    "start-element-guid",
  ],
  "connectors": Array [],
  "elements": Object {
    "start-element-guid": Object {
      "config": Object {
        "isHighlighted": false,
        "isSelectable": true,
        "isSelected": false,
      },
      "connectorCount": 0,
      "description": "",
      "elementType": "START_ELEMENT",
      "filterType": "all",
      "filters": Array [
        Object {
          "leftHandSide": "",
          "operator": "",
          "rightHandSide": "",
          "rightHandSideDataType": "",
          "rowIndex": "cda83bda-2840-470c-aa57-1a8cfedee523",
        },
      ],
      "guid": "start-element-guid",
      "isCanvasElement": true,
      "locationX": 500,
      "locationY": 120,
      "maxConnections": 1,
      "object": "",
      "objectIndex": "12c23e8e-4104-40ba-86e0-feb73e787792",
      "triggerType": "None",
    },
  },
}
`;
